### PR TITLE
[feature](merge-cloud) Enable write into cache when uploading file to s3 using s3 file writer

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

Issue Number: close #xxx
This pull request introduces significant enhancements to Apache Doris, specifically focusing on the S3 file writer and S3 file buffer pool, as outlined in issue #23953. In essence, it empowers the ability to efficiently store data in a file cache during the process of uploading data to S3 using the S3 file writer.

Notably, this PR provides users with the flexibility to control whether this caching process occurs synchronously or asynchronously through the configuration parameter `enable_flush_file_cache_async`. Also user can control whether writing into file cache or not using `enable_file_cache` config in BE.

Additionally, this PR includes improvements to the FileCache system in advance of the `BlockCache` merge, such as proactively modifying the cache type in preparation for the BlockCache merge, ensuring seamless compatibility and optimal functionality for the S3 file writer.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by Lchangliang** [comment]
LGTM

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by Lchangliang** [comment]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/io/cache/block/block_file_cache.cpp [hunk 1]
```diff
@@ -65,6 +65,8 @@ std::string IFileCache::cache_type_to_string(CacheType type) {
         return "_disposable";
     case CacheType::NORMAL:
         return "";
+    case CacheType::TTL:
```
Comment (platoneko) [inline be/src/io/cache/block/block_file_cache.cpp:68]: // TODO: consider `std::string_view`

File: be/src/io/cache/block/block_file_segment.cpp [hunk 2]
```diff
@@ -196,6 +196,36 @@ Status FileBlock::read_at(Slice buffer, size_t read_offset) {
...
+    std::unique_lock segment_lock(_mutex);
+    if (new_type == _cache_type) {
+        return true;
+    }
+    if (_download_state == State::DOWNLOADED) {
+        std::error_code ec;
+        std::filesystem::rename(get_path_in_local_cache(),
+                                _cache->get_path_in_local_cache(key(), offset(), new_type), ec);
+        if (ec) {
+            LOG(ERROR) << ec.message();
```
Comment (platoneko) [inline be/src/io/cache/block/block_file_segment.cpp:209]: Add some context in log

File: be/src/io/fs/s3_file_writer.cpp [hunk 3]
```diff
@@ -85,11 +88,18 @@ S3FileWriter::S3FileWriter(std::string key, std::shared_ptr<S3FileSystem> fs,
         : FileWriter(fmt::format("s3://{}/{}", fs->s3_conf().bucket, key), fs),
           _bucket(fs->s3_conf().bucket),
           _key(std::move(key)),
-          _client(fs->get_client()) {
+          _client(fs->get_client()),
+          _expiration_time(opts ? opts->file_cache_expiration : 0),
+          _is_cold_data(opts ? opts->is_cold_data : true),
+          _disable_file_cache(!opts ? false : !opts->write_file_cache) {
```
Comment (platoneko) [inline be/src/io/fs/s3_file_writer.cpp:94]: Rename to `_write_file_cache`

File: be/src/service/doris_main.cpp [hunk 4]
```diff
@@ -60,7 +60,7 @@
 #include "common/signal_handler.h"
 #include "common/status.h"
 #include "io/cache/block/block_file_cache_factory.h"
-#include "io/fs/s3_file_write_bufferpool.h"
```
Comment (platoneko) [inline be/src/service/doris_main.cpp:63]: Why need to include this?

**Review by platoneko — CHANGES_REQUESTED** [review]
Pls resolve below comment.

File: be/src/service/doris_main.cpp (hunk 4 above)
Comment (ByteYue) [reply be/src/service/doris_main.cpp:63]: It was the previous implementation for eager buffer pool memory allocation. It's been removed now

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

File: be/src/io/io_common.h [hunk 5]
```diff
@@ -52,8 +52,10 @@ struct IOContext {
     // FIXME(plat1ko): Seems `is_disposable` can be inferred from the `reader_type`?
     bool is_disposable = false;
     bool is_index_data = false;
-    bool read_file_cache = true;
+    // TODO(lightman): use following member variables to control file cache
+    bool is_persistent = false;
     int64_t expiration_time = 0;
```
Comment (platoneko) [inline be/src/io/io_common.h:57]: 别改顺序，节省点struct size

File: be/src/io/fs/s3_file_writer.cpp [hunk 6]
```diff
@@ -85,11 +88,18 @@ S3FileWriter::S3FileWriter(std::string key, std::shared_ptr<S3FileSystem> fs,
...
-          _client(fs->get_client()) {
+          _client(fs->get_client()),
+          _expiration_time(opts ? opts->file_cache_expiration : 0),
+          _is_cold_data(opts ? opts->is_cold_data : true),
+          _write_file_cache(!opts ? false : !opts->write_file_cache) {
     s3_file_writer_total << 1;
     s3_file_being_written << 1;
 
     Aws::Http::SetCompliantRfc3986Encoding(true);
+    if (config::enable_file_cache && !_write_file_cache) {
```
Comment (platoneko) [inline be/src/io/fs/s3_file_writer.cpp:99]: May be .. && _write_file_cache ?

**Comment by ByteYue** [comment]
run buildall

File: be/src/io/fs/s3_file_writer.cpp [hunk 7]
```diff
@@ -85,11 +88,18 @@ S3FileWriter::S3FileWriter(std::string key, std::shared_ptr<S3FileSystem> fs,
         : FileWriter(fmt::format("s3://{}/{}", fs->s3_conf().bucket, key), fs),
           _bucket(fs->s3_conf().bucket),
           _key(std::move(key)),
-          _client(fs->get_client()) {
+          _client(fs->get_client()),
+          _expiration_time(opts ? opts->file_cache_expiration : 0),
+          _is_cold_data(opts ? opts->is_cold_data : true),
+          _write_file_cache(!opts ? false : !opts->write_file_cache) {
```
Comment (platoneko) [inline be/src/io/fs/s3_file_writer.cpp:94]:
```suggestion
          _write_file_cache(opts ? opts->write_file_cache : false) {
```

File: be/src/io/fs/s3_file_writer.cpp [hunk 8]
```diff
@@ -205,30 +221,58 @@ Status S3FileWriter::close() {
...
+                                this->_st = std::move(s);
+                            }
+                            // After the signal, there is a scenario where the previous invocation of _wait_until_finish
+                            // returns to the caller, and subsequently, the S3 file writer is destructed.
+                            // This means that accessing _failed afterwards would result in a heap use after free vulnerability.
+                            _countdown_event.signal();
+                            return ret;
+                        })
+                        .set_is_cancelled([this]() { return _failed.load(); });
+                if (!_write_file_cache) {
```
Comment (platoneko) [inline be/src/io/fs/s3_file_writer.cpp:258]:
```suggestion
                if (_write_file_cache) {
```

File: be/src/io/cache/block/block_file_cache.cpp [hunk 9]
```diff
@@ -65,6 +65,8 @@ std::string IFileCache::cache_type_to_string(CacheType type) {
...
+        return "_ttl";
```
Comment (platoneko) [inline be/src/io/cache/block/block_file_cache.cpp:69]: return `std::string_view`

File: be/src/io/fs/s3_file_writer.h [hunk 10]
```diff
@@ -68,20 +69,26 @@ class S3FileWriter final : public FileWriter {
...
     std::string _upload_id;
+    size_t _index_offset {0};
 
     // Current Part Num for CompletedPart
     int _cur_part_num = 1;
     std::mutex _completed_lock;
     std::vector<std::unique_ptr<Aws::S3::Model::CompletedPart>> _completed_parts;
 
+    IFileCache::Key _cache_key;
+    IFileCache* _cache;
```
Comment (platoneko) [inline be/src/io/fs/s3_file_writer.h:80]: 建议初始化一下

**Review by platoneko — CHANGES_REQUESTED** [review]

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by platoneko — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**